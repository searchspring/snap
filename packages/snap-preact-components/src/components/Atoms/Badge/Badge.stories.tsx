/** @jsx h */
import { h } from 'preact';

import { ArgsTable, PRIMARY_STORY } from '@storybook/addon-docs/blocks';
import { componentArgs } from '../../../utilities';
import { Badge } from './Badge';

// @ts-ignore
import Readme from '../Badge/readme.md';

export default {
	title: `Atoms/Badge`,
	component: Badge,
	parameters: {
		docs: {
			page: () => (
				<div>
					<Readme />
					<ArgsTable story={PRIMARY_STORY} />
				</div>
			),
		},
	},
	decorators: [
		(Story) => (
			<div
				style={{
					margin: '1em',
					width: '220px',
					height: '300px',
					position: 'relative',
					border: '1px solid lightgrey',
				}}
			>
				<Story height="200px" />
			</div>
		),
	],
	argTypes: {
		content: {
			description: 'Content to be displayed in badge',
			table: {
				type: {
					summary: 'string',
				},
			},
			control: { type: 'text' },
		},
		children: {
			description: 'Content to be displayed in badge using children',
			table: {
				type: {
					summary: 'string, JSX',
				},
			},
		},
		position: {
			description: 'Position of badge',
			table: {
				type: {
					summary: 'object',
				},
				defaultValue: { summary: '{ top: 0, left: 0 }' },
			},
			control: { type: 'object' },
		},
		...componentArgs,
	},
};

const Template = (args) => <Badge {...args} />;

export const BadgeWithContent = Template.bind({});
BadgeWithContent.args = {
	content: 'pink',
	position: {
		top: 0,
		right: 0,
	},
};

export const BadgeWithChildren = (args) => (
	<Badge {...args}>
		<img src="//cdn.searchspring.net/ajax_search/img/star-badge-new-blue.png" />
	</Badge>
);
