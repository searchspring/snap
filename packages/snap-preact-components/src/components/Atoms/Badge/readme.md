## Badge

Renders an absolute-positioned badge. It is expected that the parent element contains `position: relative`.

## Usage

### content
The `content` prop specifies the badge.

```jsx
<div style="position: relative;">
    <Badge content="Sale" />
</div>
```

Or alternatively using children:

```jsx    
<div style="position: relative;">
    <Badge>Sale</Badge>
</div>
```

### position
The `position` prop specifies an object with CSS `top`, `bottom`, `left`, and `right` attributes. The default position is top left `{ top: 0, left: 0 }`.

In this example, the badge will be 2px from the top and 2px from the right:

```jsx    
<Badge position={{ "top": 2, "right": 2 }}>Sale</Badge>
```
