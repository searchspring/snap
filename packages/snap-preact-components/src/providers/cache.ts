export { CacheProvider } from '@emotion/react';
import createCache from '@emotion/cache';

export const cache = createCache({
	key: 'ss',
	prepend: true,
});
