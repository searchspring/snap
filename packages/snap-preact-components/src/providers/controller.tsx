import { jsx } from '@emotion/react';
import preact, { h, createContext, ComponentChildren, ComponentType } from 'preact';
import { useContext } from 'preact/hooks';
import type { AbstractController } from '@searchspring/snap-controller';

const ControllerContext = createContext<null | AbstractController>(null);

export const ControllerProvider = ({ children, controller }: { children: ComponentChildren; controller: AbstractController }) => {
	return <ControllerContext.Provider value={controller}>{children}</ControllerContext.Provider>;
};

export const useController = () => useContext(ControllerContext);

export const withController = (Component: ComponentType) => (props: any) => {
	return <Component controller={useController()} {...props} />;
};
