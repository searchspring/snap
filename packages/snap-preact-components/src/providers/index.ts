export * from './store';
export * from './theme';
