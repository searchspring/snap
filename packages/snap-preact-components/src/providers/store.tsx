import preact, { h, createContext, ComponentChildren, ComponentType } from 'preact';
import { useContext } from 'preact/hooks';
import type { AbstractStore } from '@searchspring/snap-store-mobx';

const StoreContext = createContext<null | AbstractStore>(null);

export const StoreProvider = ({ children, store }: { children: ComponentChildren; store: AbstractStore }) => {
	return <StoreContext.Provider value={store}>{children}</StoreContext.Provider>;
};

export const useStore = () => useContext(StoreContext);

export function withStore<C extends ComponentType>(Component: C): C {
	return ((props: any) => <StoreContext.Consumer>{(store) => <Component {...props} store={store} />}</StoreContext.Consumer>) as C;
}
