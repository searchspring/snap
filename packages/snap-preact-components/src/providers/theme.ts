import { ThemeProvider as EmotionTheme } from '@emotion/react';
export { css, useTheme, withTheme } from '@emotion/react';

interface ThemeProviderProps {
	theme: Partial<Theme> | ((outerTheme: Theme) => Theme);
	children?: any;
}
interface ThemeProvider {
	(props: ThemeProviderProps): React.ReactElement;
}
export const ThemeProvider: ThemeProvider = EmotionTheme;

export const defaultTheme: Theme = {
	colors: {
		primary: '#3A23AD',
		secondary: '#00cee1',
		hover: '#f8f6fd',
		text: {
			secondary: '#ffffff',
		},
		message: {
			error: '#cc1212',
			warning: '#ecaa15',
			info: '#4c3ce2',
		},
	},
	components: {},
};
export interface Theme {
	variables?: {
		[key: string]: unknown;
	};
	colors?: {
		primary: string;
		secondary: string;
		hover?: string;
		text?: {
			secondary: string;
		};
		message?: {
			error: string;
			warning: string;
			info: string;
		};
	};
	components?: any;
	namedComponents?: any;
}
