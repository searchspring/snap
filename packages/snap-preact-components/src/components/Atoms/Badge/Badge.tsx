/** @jsx jsx */
import { ComponentChildren, h } from 'preact';

import { jsx, css } from '@emotion/react';
import classnames from 'classnames';
import { observer } from 'mobx-react-lite';

import { Theme, useTheme, CacheProvider } from '../../../providers';
import { ComponentProps, StylingCSS } from '../../../types';
import { mergeProps } from '../../../utilities';

const CSS = {
	badge: ({ position }: BadgeProps) =>
		css({
			display: 'inline-block',
			position: 'absolute',
			...position,
		}),
};

export const Badge = observer((properties: BadgeProps): JSX.Element => {
	const globalTheme: Theme = useTheme();
	const theme = { ...globalTheme, ...properties.theme };
	const defaultProps: Partial<BadgeProps> = {
		position: {
			top: 0,
			left: 0,
		},
	};

	const props = mergeProps('badge', globalTheme, defaultProps, properties);

	const { content, title, children, disableStyles, className, style, styleScript } = props;

	const styling: { css?: StylingCSS } = {};
	const stylingProps = { ...props, theme };

	if (styleScript) {
		styling.css = [styleScript(stylingProps), style];
	} else if (!disableStyles) {
		styling.css = [CSS.badge(stylingProps), style];
	} else if (style) {
		styling.css = [style];
	}

	return (
		<CacheProvider>
			<div {...styling} className={classnames('ss__badge', className)} title={title}>
				{content || children}
			</div>
		</CacheProvider>
	);
});

export interface BadgeProps extends ComponentProps {
	content?: any;
	children?: ComponentChildren;
	position?: {
		top?: string | number;
		right?: string | number;
		bottom?: string | number;
		left?: string | number;
	};
	title?: string;
}
