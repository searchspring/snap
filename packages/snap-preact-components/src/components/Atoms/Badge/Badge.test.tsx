import { h } from 'preact';

import { render } from '@testing-library/preact';
import { ThemeProvider } from '../../../providers/theme';

import { Badge } from './Badge';

const CONTENT = 'sale';
const CLASSNAME = 'sale-badge';
const CHILDREN = <div>{CONTENT}</div>;

// TODO: Add test to fully cover position prop

describe('Badge Component', () => {
	it('positions badge based on prop', () => {
		const position = { right: 0 };
		const rendered = render(<Badge position={position}>{CHILDREN}</Badge>);
		const badge = rendered.container.querySelector('.ss-badge');
		let styles = getComputedStyle(badge);

		expect(styles.right).toEqual('0px');

		const newPosition = { left: 0 };
		rendered.rerender(<Badge position={newPosition}>{CHILDREN}</Badge>);
		styles = getComputedStyle(badge);

		expect(styles.left).toEqual('0px');
	});

	describe('Badge with content', () => {
		let rendered;
		beforeEach(() => {
			rendered = render(<Badge className={CLASSNAME} content={CONTENT} />);
		});

		it('renders Badge', () => {
			const BadgeElement = rendered.getByText(CONTENT);

			expect(BadgeElement).toBeInTheDocument();
		});

		it('Badge has correct classes', () => {
			const badge = rendered.container.querySelector('.ss-badge');
			expect(badge).toBeInTheDocument();
			expect(badge.classList.length).toBe(3);
			expect(badge.classList[0]).toMatch(/^ss-badge/);
			expect(badge).toHaveClass(CLASSNAME);
		});
	});

	describe('Badge with children', () => {
		let badge;
		beforeEach(() => {
			const position = { right: 0 };
			badge = render(<Badge position={position}>{CHILDREN}</Badge>);
		});

		it('renders Badge', () => {
			const BadgeElement = badge.getByText(CONTENT);

			expect(BadgeElement).toBeInTheDocument();
		});

		it('Badge has correct number of classes', () => {
			const BadgeElement = badge.getByText(CONTENT).parentElement;

			expect(BadgeElement.classList.length).toBe(2);
			expect(BadgeElement.classList[0]).toMatch(/^ss-badge/);
		});
	});

	describe('Badge with disableStyles', () => {
		let badge;
		beforeEach(() => {
			badge = render(<Badge content={CONTENT} disableStyles />);
		});

		it('renders Badge', () => {
			const BadgeElement = badge.getByText(CONTENT);

			expect(BadgeElement).toBeInTheDocument();
		});

		it('Badge has correct number of classes', () => {
			const BadgeElement = badge.getByText(CONTENT);

			expect(BadgeElement.classList.length).toBe(2);
			expect(BadgeElement.classList[1]).toMatch(/^css-0/);
		});
	});

	describe('Badge theming works', () => {
		it('is themeable with ThemeProvider', () => {
			const globalTheme = {
				components: {
					badge: {
						className: 'classy',
					},
				},
			};
			const position = { right: 0 };
			const rendered = render(
				<ThemeProvider theme={globalTheme}>
					<Badge position={position}>{CHILDREN}</Badge>
				</ThemeProvider>
			);
			const element = rendered.container.querySelector('.ss-badge');
			expect(element).toBeInTheDocument();
			expect(element).toHaveClass(globalTheme.components.badge.className);
		});

		it('is themeable with theme prop', () => {
			const propTheme = {
				components: {
					badge: {
						className: 'classy',
					},
				},
			};
			const position = { right: 0 };
			const rendered = render(
				<Badge position={position} theme={propTheme}>
					{CHILDREN}
				</Badge>
			);
			const element = rendered.container.querySelector('.ss-badge');
			expect(element).toBeInTheDocument();
			expect(element).toHaveClass(propTheme.components.badge.className);
		});

		it('is theme prop overrides ThemeProvider', () => {
			const globalTheme = {
				components: {
					badge: {
						className: 'classy',
						content: 'oooo',
					},
				},
			};
			const propTheme = {
				components: {
					badge: {
						className: 'classier',
						content: 'ahhhh',
					},
				},
			};
			const position = { right: 0 };
			const rendered = render(
				<ThemeProvider theme={globalTheme}>
					<Badge position={position} theme={propTheme}>
						{CHILDREN}
					</Badge>
				</ThemeProvider>
			);

			const element = rendered.container.querySelector('.ss-badge');
			expect(element).toBeInTheDocument();
			expect(element).toHaveClass(propTheme.components.badge.className);
			expect(element).not.toHaveClass(globalTheme.components.badge.className);
			expect(element).toHaveTextContent(propTheme.components.badge.content);
			expect(element).not.toHaveTextContent(globalTheme.components.badge.content);
		});
	});
});
