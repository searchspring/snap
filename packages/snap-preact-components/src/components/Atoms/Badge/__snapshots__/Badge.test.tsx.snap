// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge Component Badge theming works is theme prop overrides ThemeProvider 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: inline-block;
  position: absolute;
  right: 0;
}

<div
    class="ss__badge classier emotion-0"
  >
    ahhhh
  </div>
</DocumentFragment>
`;

exports[`Badge Component Badge theming works is themeable with ThemeProvider 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: inline-block;
  position: absolute;
  right: 0;
}

<div
    class="ss__badge classy emotion-0"
  >
    <div>
      sale
    </div>
  </div>
</DocumentFragment>
`;

exports[`Badge Component Badge theming works is themeable with theme prop 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: inline-block;
  position: absolute;
  right: 0;
}

<div
    class="ss__badge classy emotion-0"
  >
    <div>
      sale
    </div>
  </div>
</DocumentFragment>
`;

exports[`Badge Component Badge with children Badge has correct number of classes 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: inline-block;
  position: absolute;
  right: 0;
}

<div
    class="ss__badge emotion-0"
  >
    <div>
      sale
    </div>
  </div>
</DocumentFragment>
`;

exports[`Badge Component Badge with children renders Badge 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: inline-block;
  position: absolute;
  right: 0;
}

<div
    class="ss__badge emotion-0"
  >
    <div>
      sale
    </div>
  </div>
</DocumentFragment>
`;

exports[`Badge Component Badge with content Badge has correct classes 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: inline-block;
  position: absolute;
  top: 0;
  left: 0;
}

<div
    class="ss__badge sale-badge emotion-0"
  >
    sale
  </div>
</DocumentFragment>
`;

exports[`Badge Component Badge with content renders Badge 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: inline-block;
  position: absolute;
  top: 0;
  left: 0;
}

<div
    class="ss__badge sale-badge emotion-0"
  >
    sale
  </div>
</DocumentFragment>
`;

exports[`Badge Component Badge with disableStyles Badge has correct number of classes 1`] = `
<DocumentFragment>
  <div
    class="ss__badge"
  >
    sale
  </div>
</DocumentFragment>
`;

exports[`Badge Component Badge with disableStyles renders Badge 1`] = `
<DocumentFragment>
  <div
    class="ss__badge"
  >
    sale
  </div>
</DocumentFragment>
`;

exports[`Badge Component positions badge based on prop 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: inline-block;
  position: absolute;
  right: 0;
}

<div
    class="ss__badge emotion-0"
  >
    <div>
      sale
    </div>
  </div>
</DocumentFragment>
`;

exports[`Badge Component positions badge based on prop 2`] = `
<DocumentFragment>
  .emotion-0 {
  display: inline-block;
  position: absolute;
  left: 0;
}

<div
    class="ss__badge emotion-0"
  >
    <div>
      sale
    </div>
  </div>
</DocumentFragment>
`;

exports[`Badge Component positions badge based on prop 3`] = `
<DocumentFragment>
  .emotion-0 {
  display: inline-block;
  position: absolute;
  bottom: 0;
}

<div
    class="ss__badge emotion-0"
  >
    <div>
      sale
    </div>
  </div>
</DocumentFragment>
`;

exports[`Badge Component positions badge based on prop 4`] = `
<DocumentFragment>
  .emotion-0 {
  display: inline-block;
  position: absolute;
  top: 0;
}

<div
    class="ss__badge emotion-0"
  >
    <div>
      sale
    </div>
  </div>
</DocumentFragment>
`;
